File: Sorting/CountingSort/countingSort.py
def printArr(data):
    for i in range(len(data)):
        print(data[i],end="....")
    print("\n")

def countingSort(data):
    k = max(data)
    output = [0]*len(data)        
    count = [0]*(k+1)
    
    for i in range(0,len(data)):
        count[data[i]] += 1
        
    for i in range(1,k+1):
        count[i] += count[i-1]
        
    i = len(data) - 1
    
    while i>=0:
        output[count[data[i]]-1] = data[i]
        count[data[i]] -= 1
        i -= 1
        
    for i in range(0,len(data)):
        data[i] = output[i]
    

if __name__ == '__main__':
    data = [4,2,5,2,4,2,0,9,8,7,6,5,4,3,2,1,3,5,7,3]
    printArr(data)
    countingSort(data)
    printArr(data)    

File: Sorting/CountingSort/countingSort.java


class countingSort
{
    static void printArr(int[] arr)
    {
        for(int i=0;i<arr.length;i++)
        {
            System.out.print(arr[i]+"\t");
        }
        System.out.println("\n");
    }
    static void sort(int[] arr,int len)
    {
        int k = arr[0];
        for(int i=1;i<arr.length;i++)
        {
            k=Math.max(k,arr[i]);
        }
        int output[] = new int[len];
        int count[] = new int[k+1];
        for(int i=0;i<k;i++)
        {
            count[i]=0;
        }
        for(int i=0;i<len;i++)
        {
            count[arr[i]]++;
        }
        for(int i=1;i<=k;i++)
        {
            count[i] += count[i-1];
        }
        for(int i=len-1;i>=0;i--)
        {
            output[count[arr[i]]-1] = arr[i];
            count[arr[i]]--;
        }
        for(int i=0;i<len;i++)
        {
            arr[i]=output[i];
        }
    }
    public static void main(String[] args)
    {
        int[] data = {1,4,2,4,5,2,5,2,6,7,4,7,4,8,9,3,1,5,0,4,2};
        int len = data.length;
        printArr(data);
        sort(data,len);
        printArr(data);
    }
}